src/test/unit/mcmc/chainset_test.cpp

```diff
@@ -172,163 +142,70 @@ TEST_F(McmcChains, eight_schools_samples) {
-  Eigen::VectorXd probs(3);
-  probs << 0.05, 0.5, 0.95;
-
-  for (size_t i = 0; i < 10; ++i) {
-    auto mean = chain_2.mean(i + 7);
-    EXPECT_NEAR(mean, s8_mean(i), 0.05);
-    auto median = chain_2.median(i + 7);
-    EXPECT_NEAR(median, s8_median(i), 0.05);
-    auto sd = chain_2.sd(i + 7);
-    EXPECT_NEAR(sd, s8_sd(i), 0.05);
-    auto mad = chain_2.max_abs_deviation(i + 7);
-    EXPECT_NEAR(mad, s8_mad(i), 0.05);
-    auto q_5 = chain_2.quantile(i + 7, 0.05);
-    EXPECT_NEAR(q_5, s8_q5(i), 0.5);
-    auto q_95 = chain_2.quantile(i + 7, 0.95);
-    EXPECT_NEAR(q_95, s8_q95(i), 0.5);
-    auto qs_5_50_95 = chain_2.quantiles(i + 7, probs);
-    EXPECT_EQ(3, qs_5_50_95.size());
-    EXPECT_NEAR(qs_5_50_95(0), s8_q5(i), 0.5);
-    EXPECT_NEAR(qs_5_50_95(1), s8_median(i), 0.05);
-    EXPECT_NEAR(qs_5_50_95(2), s8_q95(i), 0.5);
+  std::stringstream out;
+  std::vector<stan::io::stan_csv> bern_csvs(4);
+  for (size_t i = 0; i < 4; ++i) {
+    std::stringstream fname;
+    fname << "src/test/unit/analyze/mcmc/test_csv_files/bern" << (i + 1)
+          << ".csv";
+    std::ifstream bern_stream(fname.str(), std::ifstream::in);
+    stan::io::stan_csv bern_csv
+        = stan::io::stan_csv_reader::parse(bern_stream, &out);
+    bern_stream.close();
+    bern_csvs[i] = bern_csv;
   }
-}
-
-TEST_F(McmcChains, mcse) {
-  std::vector<stan::io::stan_csv> eight_schools;
-  eight_schools.push_back(eight_schools_1);
-  eight_schools.push_back(eight_schools_2);
-  stan::mcmc::chainset chain_2(eight_schools);
-  EXPECT_EQ(2, chain_2.num_chains());
-
-  // test against R implementation in pkg posterior
-  Eigen::VectorXd s8_mcse_mean(10), s8_mcse_sd(10);
-  s8_mcse_mean << 0.288379, 0.4741815, 0.2741001, 0.3294614, 0.2473758,
-      0.2665048, 0.2701363, 0.4740092, 0.3621771, 0.3832464;
-  s8_mcse_sd << 0.1841825, 0.2854258, 0.192332, 0.2919369, 0.2478025, 0.2207478,
-      0.2308452, 0.2522107, 0.2946896, 0.3184745;
-
-  for (size_t i = 0; i < 10; ++i) {
-    auto mcse_mean = chain_2.mcse_mean(i + 7);
-    EXPECT_NEAR(mcse_mean, s8_mcse_mean(i), 0.5);
-    auto mcse_sd = chain_2.mcse_sd(i + 7);
-    EXPECT_NEAR(mcse_sd, s8_mcse_sd(i), 0.7);
+  stan::mcmc::chainset bern_chains(bern_csvs);
+  EXPECT_EQ(4, bern_chains.num_chains());
+
+  Eigen::MatrixXd theta = bern_chains.samples("theta");
+  // default summary statistics - via R pkg posterior
+  double theta_mean_expect = 0.2512974105;
+  EXPECT_NEAR(theta_mean_expect, bern_chains.mean("theta"), 0.00001);
+
+  double theta_median_expect = 0.237476;
+  EXPECT_NEAR(theta_median_expect, bern_chains.median("theta"), 0.00001);
+
+  double theta_sd_expect = 0.1215466867;
+  EXPECT_NEAR(theta_sd_expect, bern_chains.sd("theta"), 0.00001);
+
+  double theta_mad_expect = 0.1230906411;
+  EXPECT_NEAR(theta_mad_expect, bern_chains.max_abs_deviation("theta"),
+              0.00001);
+
+  double theta_mcse_mean_expect = 0.0032339916;
+  EXPECT_NEAR(theta_mcse_mean_expect, bern_chains.mcse_mean("theta"), 0.0001);
```

Comment:
It is still incongruous to put 10 digits in the "expected" answer but only test four digits based on tolerances. 

Adding more precision shows that these two computed values aren't even the same out to 4 digits if you _round_ (expected rounds to .0032, computed rounds to .0033), only if you truncate. And, it still isn't precise enough to differentiate between using the sample variance and incorrectly using population variance

Reply:
I want a test to do at least one of two things:

1. Convince me we're calculating the correct quantity by re-creating (to high precision) the result of another package
2. Provide some reasonable certainty for future refactorings that they _continue_ to calculate the correct value

I would say right now it does neither. We aren't getting the same answer as R, nor are we testing tightly enough that we would catch simple bugs during a future code change.

4000 draws results in a limited amount of resolution that is useful for an end user, sure, but this is a deterministic calculation. For a given sample, if we were computing in infinite precision, there would be exactly one answer that properly implementing the algorithm would compute. We're not in infinite precision, of course, but I would expect R and C++ to get much closer than this. If they don't, then I think it is important to understand why so that we can either make them agree, or document why they do not

Reply:
> And, it still isn't precise enough to differentiate between using the sample variance and incorrectly using population variance

a review of the code will demonstrate that we are using sample variance, not population variance.

Reply:
I have changed the `rhat_basic_test` and `ess_basic_test` so that they check that the previous implementation of rhat and ess provides the same answer to 9 digits of accuracy.

the implementation of `mcse_mean` and `mcse_sd` match the implementations in R's posterior package.  they depend on the implementation of `ess` being correct, as demonstrated by the `ess_basic_test`.

Reply:
> I want a test to do at least one of two things:
> 
> 1. Convince me we're calculating the correct quantity by re-creating (to high precision) the result of another package
> 2. Provide some reasonable certainty for future refactorings that they _continue_ to calculate the correct value

see above - the calculations of rhat and ess are the same to 9 digits.